noise/README.org

```diff
@@ -0,0 +1,59 @@
+The Noise plugin allows sending and receiving private messages through the
+Lightning Network. It is implemented on top to c-lightning's ~createonion~ and
+~sendonion~ RPC methods that allow delivering custom payloads to a specific
+node, as well as the ~htlc_accepted~ hook which can be used to extract the
+message from the onion payload.
+
+You can send a message using the following RPC method:
+
+#+BEGIN_SRC bash
+lightning-cli sendmsg 02a5deaa47804c518bb4a1c6f04a85b92b796516bd32c4114a51b00d73e251f999 "Hello world 👋"
+#+END_SRC
+
+In addition a message can also be accompanied by a payment (using the
+~keysend~ protocol draft) by specifying an amount of millisatoshis as the last
+argument:
+
+#+BEGIN_SRC bash
+lightning-cli sendmsg 02a5deaa47804c518bb4a1c6f04a85b92b796516bd32c4114a51b00d73e251f999 "Here's my rent" 31337
+#+END_SRC
+
+You can read the last message received using the following command:
+
+#+BEGIN_SRC bash
+lightning-cli recvmsg last_id
+#+END_SRC
+
+The ~last_id~ indicates the last message we read, so we can retrieve each message
+individually. If you'd just like to wait for the next message you can use a
+~last_id~ or ~-1~.
+* Todo
+
+- [ ] Persist messages across restarts
+- [ ] Use ~rpc_command~ to intercept any payment listing and add the keysend
+  payments to it.
+
+* Protocol
+The protocol was heavily inspired by the [[https://github.com/joostjager/whatsat#protocol][WhatSat protocol]]:
+
+| record type | length (bytes) | value                                                           |
+|-------------+----------------+-----------------------------------------------------------------|
+|  5482373484 |             32 | key send preimage                                               |
+|    34349334 |       variable | chat message                                                    |
+|    34349335 |             65 | compressed signature + recovery id                              |
+|    34349339 |             33 | sender pubkey                                                   |
+|    34349343 |              8 | timestamp in nano seconds since unix epoch (big endian encoded) |
```

Comment:
Why does the first record type has a value that higher from the four following types ?

Reply:
The choice was done by @joostjager, but if we ever standardize this we'd be chosing smaller types, for the sake of conserving space in the onion.